### PR TITLE
[agent] Close iOS systeam dialogs in health check

### PR DESCRIPTION
**Pull request by digiter**
- The dialogs often cause test flakiness and performance regressions, therefore we'd like to close them before every test execution.
- This change will run the single XCUITest in the https://github.com/digiter/infra-dialog repo. The test takes about 16 secs to build and run.
- I have merged the infra-dialog repo with cocoon in https://github.com/flutter/cocoon/pull/868

Bug: https://github.com/flutter/flutter/issues/54792

### CODE REVIEW

File: agent/lib/src/health.dart [hunk 1]
```diff
@@ -142,3 +144,42 @@ Future<HealthCheckResult> removeCachedData(
...
+  List<Device> ds = await discovery.discoverDevices();
+  for (Device d in ds) {
+    if (!(d is IosDevice)) {
+      continue;
+    }
+    // Checks out the infra-dialog repo.
+    // TODO(wutong): merge infra-dialog with cocoon to get rid of this checkout.
+    // https://github.com/flutter/flutter/issues/54792
+    if (dialogDir == null) {
+      dialogDir = await getInfraDialog();
```
Comment (godofredoc) [inline agent/lib/src/health.dart:166]: What is an infra-dialog?

File: agent/lib/src/utils.dart [hunk 2]
```diff
@@ -518,6 +518,21 @@ Future<Null> getFlutter(String revision) async {
   rm(_installationLock);
 }
 
+/// Checks out the infra-dialog repo and updates it.
+Future<Directory> getInfraDialog() async {
+  var repoPath = path.join(config.flutterDirectory.parent.path, "infra-dialog");
+  Directory repoDir = Directory(repoPath);
+  if (!exists(repoDir)) {
+    await inDirectory(config.flutterDirectory.parent, () async {
+      await exec('git', ['clone', 'https://github.com/digiter/infra-dialog.git']);
```
Comment (godofredoc) [inline agent/lib/src/utils.dart:527]: How difficult would it be to merge the code in cocoon as part of this PR?

File: agent/lib/src/utils.dart [hunk 3]
```diff
@@ -633,3 +648,14 @@ Future<void> killAllRunningProcessesOnWindows(String processName) async {
 /// futures are intentionally not awaited. This function may be used to ignore a
 /// particular future. It silences the unawaited_futures lint.
 void unawaited(Future<void> future) {}
+
+Future<Null> unlockKeyChain() async {
```
Comment (godofredoc) [inline agent/lib/src/utils.dart:652]: Docs?

File: agent/lib/src/health.dart [hunk 4]
```diff
@@ -142,3 +144,42 @@ Future<HealthCheckResult> removeCachedData(
...
+Future<HealthCheckResult> closeIosDialog(
+    {ProcessManager pm = const LocalProcessManager(), DeviceDiscovery discovery, Directory dialogDir}) async {
+  if (discovery == null) {
+    discovery = devices;
+  }
+  List<Device> ds = await discovery.discoverDevices();
+  for (Device d in ds) {
+    if (!(d is IosDevice)) {
+      continue;
+    }
```
Comment (keyonghan) [inline agent/lib/src/health.dart:161]: How about using `Device d = ds.where((Device d) => d is IosDevice)` to remove the for loop?

File: agent/lib/src/health.dart [hunk 5]
```diff
@@ -142,3 +144,42 @@ Future<HealthCheckResult> removeCachedData(
...
+/// Closes system dialogs on iOS, e.g. the one about new system update.
+///
+/// The dialogs often cause test flakiness and performance regressions.
+@visibleForTesting
+Future<HealthCheckResult> closeIosDialog(
+    {ProcessManager pm = const LocalProcessManager(), DeviceDiscovery discovery, Directory dialogDir}) async {
+  if (discovery == null) {
+    discovery = devices;
+  }
+  List<Device> ds = await discovery.discoverDevices();
```
Comment (keyonghan) [inline agent/lib/src/health.dart:157]: Nit: full name?

File: agent/lib/src/utils.dart (hunk 3 above)
Comment (digiter) [reply agent/lib/src/utils.dart:645]: Added.

**Pull request merged**